### ISSUE TITLE
Slow `jit` compilation time compared to `jax.experimental.ode.odeint`

### ISSUE THREAD

**Issue by dominicrufa** [patrick-kidger/diffrax#94]
hi @patrick-kidger,
big fan of `diffrax`!

I've been playing around with some of the functionality you have in this repository and comparing it with the ode-solver in `jax`. The one pain point i noticed is that there seems to be a relatively slow `jit` compilation time, particularly when I try to `jit` the `grad` of a simple loss function containing `diffeqsolve`. I was wondering if this is an error on my part (perhaps I botched the `diffrax` implementation) or if there is yet to be some optimization. The demonstration is below:
 
```
from jax.config import config
config.update("jax_enable_x64", True)
config.update("jax_debug_nans", True) 
config.parse_flags_with_absl()
import jax
import jax.numpy as jnp
from jax import random
import numpy as np
from functools import partial
import haiku as hk

def exact_kinematic_aug_diff_f(t, y, args_tuple):
    """
    """
    _y, _, _ = y
    _params, _key, diff_f = args_tuple
    aug_diff_fn = lambda __y : diff_f(t, __y, (_params,))
    _f, s, t = aug_diff_fn(_y)
    r = jnp.sum(t)
    return _f, r, 0.

def exact_kinematic_odeint_diff_f(y, t, params, canonical_diff_fn):
    run_y = y[0]
    _f, s, t = canonical_diff_fn(t, run_y, (params,))
    return _f, jnp.sum(s), 0.

class TestMLP(hk.Module):
    def __init__(self, num_particles, name=None):
        super().__init__(name=None)
        self._mlp = hk.nets.MLP([8,8,8,8,num_particles*12])
        self._num_particles=num_particles
    def __call__(self, t, y):
        in_y = (y + t).flatten()
        outter = self._mlp(in_y).reshape((4, self._num_particles, 3))
        return outter[:2], outter[2], outter[3]

def test(num_particles):
    import functools
    from jax.experimental.ode import odeint
    import diffrax
    
    #generate positions/velocities
    small_positions = jax.random.normal(jax.random.PRNGKey(261), shape=(num_particles,3))
    small_velocities = jax.random.normal(jax.random.PRNGKey(235), shape=(num_particles,3))
    small_positions_and_velocities = jnp.vstack([small_positions[jnp.newaxis, ...], small_velocities[jnp.newaxis, ...]])
    
    # make module kwargs
    VectorMLP_kwargs = {'num_particles': num_particles}
    
    # make module function
    def _diff_f_wrapper(t, y):
        diff_f = TestMLP(**VectorMLP_kwargs)
        return diff_f(t, y)
    
    diff_f_init, diff_f_apply = hk.without_apply_rng(hk.transform(_diff_f_wrapper))
    init_params = diff_f_init(jax.random.PRNGKey(36), 0., small_positions_and_velocities)
    canonicalized_diff_f_fn = lambda _t, _y, _args_tuple : diff_f_apply(_args_tuple[0], _t, _y)
    
    # make the augmented functions
    odeint_aug_diff_func = functools.partial(exact_kinematic_odeint_diff_f, canonical_diff_fn=canonicalized_diff_f_fn)
    diffeqsolve_aug_diff_func = exact_kinematic_aug_diff_f
    
    # odeint solver
    def odeint_solver(_parameters, _init_y, _key):
        aug_init_y = (_init_y, 0., 0.)
        outs = odeint(odeint_aug_diff_func, aug_init_y, jnp.array([0., 1.]), _parameters, rtol=1.4e-8, atol=1.4e-8)
        final_outs = (outs[0][-1], outs[1][-1], outs[2][-1])
        return final_outs
    
    def diffrax_ode_solver(_parameters, _init_y, _key):
        term=diffrax.ODETerm(diffeqsolve_aug_diff_func)
        stepsize_controller=diffrax.PIDController(rtol=1.4e-8, atol=1.4e-8)
        solver = diffrax.Dopri5()
        aug_init_y = (_init_y, 0., 0.)
        sol = diffrax.diffeqsolve(term, 
                                  solver, 
                                  t0=0., 
                                  t1=1., 
                                  dt0=1e-1, 
                                  y0=aug_init_y, 
                                  stepsize_controller=stepsize_controller, 
                                  args=(_parameters, _key, canonicalized_diff_f_fn))
        return sol.ys[0][0], sol.ys[1][0], sol.ys[2][0]
    
    @jax.jit
    def odeint_loss_fn(_params, _init_y, _key):
        ode_solution = odeint_solver(_params, _init_y, _key)
        return jnp.sum(ode_solution[1]**2)
    
    @jax.jit
    def diffrax_loss_fn(_params, _init_y, _key):
        ode_solution = diffrax_ode_solver(_params, _init_y, _key)
        return jnp.sum((ode_solution[1])**2)
    
    # test
    import time
    
    # odeint compilation time
    start_time = time.time()
    _ = jax.grad(odeint_loss_fn)(init_params, small_positions_and_velocities, jax.random.PRNGKey(34))
    end_time = time.time()
    print(f"odeint comp. time: {end_time - start_time}")
    
    # diffrax compilation time
    start_time = time.time()
    _ = jax.grad(diffrax_loss_fn)(init_params, small_positions_and_velocities, jax.random.PRNGKey(34))
    end_time = time.time()
    print(f"diffrax comp. time: {end_time - start_time}")

```

running `test(8)` gives me the following compilation time on CPU:

```
odeint comp. time: 2.5580570697784424
diffrax comp. time: 23.965799570083618
```
I noticed that if I use `diffrax.BacksolveAdjoint`, compilation time goes down to ~8 seconds, but I'm keen to avoid that method based on your docs.; also, it looks like the compilation time in `diffrax` is heavily dependent on the number of hidden layers in `TestMLP`, perhaps suggesting a non-optimal compilation in `diffrax` of `for` loops? 
Thanks!

**Comment by patrick-kidger** [comment]
Right; quite a lot to unpack here.

First of all, note that `jax.experimental.ode.odeint` uses (what we call) `BacksolveAdjoint`, so the "fair" comparison on compilation times is indeed when using that. If you like, think of the extra compilation time of `RecursiveCheckpointAdjoint` as the price you pay for more accurate gradients...

Second, whilst it doesn't affect the timing, note that your vector fields are ever-so-slightly different between odeint and Diffrax. For one you're doing `r = sum(t)` and for the other `r = sum(s)`. (Besides which, in each case the `s`/`t` output of the net is ignored, so you can afford to make your neural network a little smaller.)

Anyway, on to the problem at hand. It turns out that yup, this is because of how Diffrax compiles things, and in particular this for loop:

https://github.com/patrick-kidger/diffrax/blob/c886c218fa300ca5d7037405703c6ae568a80aab/diffrax/solver/runge_kutta.py#L283

in the internals of the solver, where the equivalent spot in `jax.experimental.ode.odeint` uses a `jax.lax.fori_loop`.

For example switching to the lower-order `solver=diffrax.Heun()` (with a shorter `for` loop) produces
```
diffrax comp. time: 2.358006715774536
```
and a scrappy implementation that switches out that `for` loop for a `lax.scan`, produces (with `solver = diffrax.Dopri5()`):
```
diffrax comp. time: 3.7588982582092285
```
And I'm pretty sure this could probably still be halved again if I tweaked the implementation a bit -- which would then bring it in-line with `jax.experimental.ode.odeint`.

However. In principle, leaving this loop unrolled should improve runtime performance. It's actually deliberately unrolled for this reason, although I never got around to testing how big the performance improvement is. Besides enabling compiler optimisations to work across loop iterations, it means that we can steer clear of serious XLA bugs like https://github.com/google/jax/issues/10197#issuecomment-1094410615, which has the potential to destroy runtime performance. This issue in fact affects precisely the kind of computation graph that a Runge--Kutta solver needs. So whilst any runtime performance loss is actually pretty minimal on this example, I'd be antsy about changing it without pretty thoroughly benchmarking it first. (Which is another point, it'd be great to improve the number of Diffrax benchmarks we have!)

So, what to do? If this is a blocker for you then I'd be happy to accept a PR that adds a toggle for converting the above for loop into a `lax.scan`, i.e. something of the form:

```python
class AbstractRungeKutta(AbstractAdaptiveSolver):
    compile_stages: bool = False

    def step(...):
        ...

        def _scan_fun(...):
            ...

        if self.compile_stages:
            ... = lax.scan(_scan_fun, ...)
        else:
            for ... in ...:
                 ... = _scan_fun(...)
```
which you can then enable via `Dopri5(compile_stages=True)`. Then it's at least an option, and if there's any energy afterwards we can look at benchmarking this and perhaps switching over to this entirely.

(Complete side note, consider looking at [Equinox](https://github.com/patrick-kidger/equinox) if you haven't already -- this is a spin-out project from Diffrax. Amongst other things it makes working with neural networks, and parameterised vector fields, a bit simpler than other options like Haiku.)

WDYT?

**Comment by dominicrufa** [comment]
> For example switching to the lower-order solver=diffrax.Heun() (with a shorter for loop) produces
> 
> diffrax comp. time: 2.358006715774536

that's interesting. using `diffrax.Heun()` above instead of `diffrax.Dopri5()` gives me ~12s compile time.  Did you just drop and replace the solver line from the original code?

> If this is a blocker for you then I'd be happy to accept a PR that adds a toggle for converting the above for loop into a lax.scan

I'm not entirely sure this is blocking yet. However, one issue I noticed trying to replace the `for` loop with a `scan` function is that you are zipping the `tableau.a_lower` and `tableau.c`. If i were to use the `scan` [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html), and replaced `xs` with `jnp.arange(len(self.tableau.a_lower))` to iterate, it will raise a [`TraceIntegerConversionError`](https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerIntegerConversionError) since the tableau is a `Tracer`. Or maybe i am implementing this incorrectly. 

Either way, i was unfamiliar with [this issue](https://github.com/google/jax/issues/10197#issuecomment-1094410615), so maybe avoiding the rewrite would be smarter?

**Comment by patrick-kidger** [comment]
The 2.35 second compile time is when switching out to `BacksolveAdjoint` as well. (For consistency with `jax.experimental.ode.odeint`.)

As for the scan: yup, you're doing something off there; `a_lower` is a tuple of numpy arrays so you should be able to take the length of it no problem.

Anyway, if you don't need the rewrite then we can leave things as they are. At some point -- when I or anyone else gets time -- I would like to revisit this so as to improve compile times further.

**Comment by jwnys** [comment]
Just want to add that this is actually problematic for my experiments. Compilation time takes forever with diffrax (i've literally never seen the end of a compilation time), but goes quite fast with jax's odeint. I should add that I need the hessian of the log probability with respect to the spatial coordinates in CNFs (for physics applications). This even happens for the smallest systems of 2 dimensions. I would really like to use diffrax together with NetKet (https://github.com/netket/netket/issues/1109), but it seems impossible to use diffrax at all to this issue.

**Comment by jwnys** [comment]
Hi Patrick, thanks for the quick reply. I tried this, but unfortunately it seems the problem remains. I will work on a MWE in the next few days.

**Comment by druidowm** [comment]
Hi Patrick,

I am also enjoying using diffrax, and am also running into jitting speed issues when using CNFs. Here is a MWE of the issue I am running into:

```
import time
from functools import partial

import jax
import jax.nn as nn
from jax import numpy as jnp
from jax import grad, jit, vmap
from jax.random import normal,PRNGKey
from jax.scipy.stats import norm

import haiku as hk

import diffrax
from diffrax import RecursiveCheckpointAdjoint,BacksolveAdjoint

from jax.experimental.ode import odeint


def exact_logp_wrapper(func):
    def logp(t, input, params):
        y, _ = input
        fn = lambda y: func(params, y)
        f, vjp_fn = jax.vjp(fn, y)
        (size,) = y.shape
        eye = jnp.eye(size)
        (dfdy,) = jax.vmap(vjp_fn)(eye)
        logp = jnp.trace(dfdy)
        return f, logp
    return logp, lambda input,t,params: logp(t,input,params)

def get_flow(input_dim, approximate_adjoint, diffrax_flow = True):
    field_init, apply = hk.without_apply_rng(
                            hk.transform(
                                lambda x: hk.nets.MLP([10, 10, 2],activation = nn.gelu)(x)
                            )
                        )

    params = field_init(PRNGKey(0), jnp.zeros((1,input_dim),float))

    if approximate_adjoint:
        adjoint = BacksolveAdjoint()
    else:
        adjoint = RecursiveCheckpointAdjoint()

    if diffrax_flow:
        vector_field_prob = exact_logp_wrapper(apply)[0]
    else:
        vector_field_prob = exact_logp_wrapper(apply)[1]
        
    @partial(vmap, in_axes=(None,0))
    def log_prob(params, y):
        if diffrax_flow:
            term = diffrax.ODETerm(vector_field_prob)
            solver = diffrax.Dopri5()
            stepsize_controller=diffrax.PIDController(rtol=1.4e-8, atol=1.4e-8)
            
            sol = diffrax.diffeqsolve(
                term, 
                solver, 
                0.0, 0.5, 
                0.05, 
                y0 = (y,0.0), 
                args = params, 
                adjoint = adjoint,
                stepsize_controller = stepsize_controller,
            )
            (y,), (log_prob,) = sol.ys
        else:
            ys, log_probs = odeint(
                vector_field_prob,(y,0.0),
                jnp.array([0.,0.5]), 
                params,
            )
            y = ys[-1]
            log_prob = log_probs[-1]

        log_prob += norm.logpdf(y).sum(0)
        return log_prob

    return params,log_prob

def train(init_fun,approximate_adjoint,diffrax_flow):
    X = normal(PRNGKey(0),(100,2))

    input_dim = X.shape[1]

    params, log_pdf = init_fun(input_dim,approximate_adjoint=approximate_adjoint, diffrax_flow = diffrax_flow)

    @jit
    @grad
    def loss(params, inputs):
        return -log_pdf(params, inputs).mean()

    for epoch in range(10):
        start_time = time.time()
        gradient = loss(params, X)
        end_time = time.time()
        print(f"Epoch {epoch}, Time {end_time-start_time}s")

train(get_flow, approximate_adjoint = False, diffrax_flow = True)
```

When I run this, I get the following output:

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Epoch 0, Time 46.312678813934326s
Epoch 1, Time 0.02420520782470703s
Epoch 2, Time 0.02737903594970703s
Epoch 3, Time 0.027439117431640625s
Epoch 4, Time 0.03911089897155762s
Epoch 5, Time 0.02908802032470703s
Epoch 6, Time 0.03005075454711914s
Epoch 7, Time 0.025400161743164062s
Epoch 8, Time 0.02555227279663086s
Epoch 9, Time 0.025499820709228516s
```

Cutting down to a MWE actually sped up the jitting quite a bit -- the original jit times for my full CNF implementation were quite a bit longer (up to 80 seconds). 

If I change the last line to `diffrax_flow=False`, I get the following output:

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Epoch 0, Time 5.133842945098877s
Epoch 1, Time 0.00861811637878418s
Epoch 2, Time 0.0065839290618896484s
Epoch 3, Time 0.00609898567199707s
Epoch 4, Time 0.006136178970336914s
Epoch 5, Time 0.006926059722900391s
Epoch 6, Time 0.00694584846496582s
Epoch 7, Time 0.007425785064697266s
Epoch 8, Time 0.008463144302368164s
Epoch 9, Time 0.006837129592895508s
```

And if I change the last line to `approximate_adjoint = True, diffrax_flow = True`, I get

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Epoch 0, Time 18.734301805496216s
Epoch 1, Time 0.006884098052978516s
Epoch 2, Time 0.005346775054931641s
Epoch 3, Time 0.0051898956298828125s
Epoch 4, Time 0.006171703338623047s
Epoch 5, Time 0.006474971771240234s
Epoch 6, Time 0.0063381195068359375s
Epoch 7, Time 0.007714986801147461s
Epoch 8, Time 0.0066030025482177734s
Epoch 9, Time 0.006442070007324219s
```

So, although the BacksolveAdjoint does speed things up significantly, there is still a fairly large speed gap. Like @dominicrufa, I am mainly interested in the RecursiveCheckpointAdjoint.

I suspect that all the autodiff in this MWE is the problem. For example, if I remove the `@grad` from `loss`, the program only takes about 3s to compile and 1s to compile with odeint.

Thank you in advance for looking into this!

Owen

**Comment by druidowm** [comment]
Hi @patrick-kidger,

I am sorry to bother you, but I am somewhat held up by the slow jit compile time. Do you mind taking a look at the above MWE?

Thank you!

Owen

**Comment by patrick-kidger** [comment]
Hey there. I'm planning on picking this up soon-ish, but if this is a blocker for you right now then I'd be happy to accept a PR resolving it. It should be essentially straightforward: as per the above discussion all that needs to happen is to convert a `for` loop into a `lax.scan`.

**Comment by druidowm** [comment]
Hi Patrick, thanks! I will try replacing the for loop, and if it works, I'll add a PR.
